calyx-py/test/correctness/queues/binheap.py

```diff
@@ -29,107 +31,369 @@ def insert_swap(prog, name, width, len, idx_w):
+
+        and_1.left = lt_1.out
+        and_1.right = lt_2.out
+
+        and_2.left = lt_3.out
+        and_2.right = lt_4.out
+
+        while_or.left = and_1.out
+        while_or.right = and_2.out
+
+    peak = extract_snd("peak", 0, ans)
+
+    pop = [
+        peak,
+        comp.decr(size),
+        set_idx_zero,
+        cb.invoke(swap, in_a=current_idx.out, in_b=size.out, ref_mem=mem),
+        #comp.mem_store_d1(mem, size.out, cb.const(64, 0), "zero_leaf"),
+        extract_current_rank,
+        find_child_idx,
+        extract_child_l_rank,
+        extract_child_r_rank,
+        # Bubble Down
+        cb.while_with(cb.CellAndGroup(while_or, current_gt_children), 
+                [
+                    cb.if_with(cb.CellAndGroup(if_or, child_l_swap), 
+                            [
+                                cb.invoke(swap, in_a=child_l_idx.out, in_b=current_idx.out, ref_mem=mem), 
+                                set_idx_child_l
+                            ],
+                            [
+                                cb.invoke(swap, in_a=child_r_idx.out, in_b=current_idx.out, ref_mem=mem), 
+                                set_idx_child_r
+                            ]),
+                    find_child_idx,
+                    extract_child_l_rank,
+                    extract_child_r_rank
+                ])
+    ] 
+
+    push = [
+        set_idx_size,
+        store_rank_and_value,
+        comp.incr(size),
         find_parent_idx,
-        read_parent,
-        read_child,
-        cb.while_with(
-            child_lt_parent,
-            [
-                cb.invoke(
-                    swap,
-                    in_a=parent_idx.out,
-                    in_b=child_idx.out,
-                    ref_mem=mem,
-                ),
-                bubble_child_idx,
-                find_parent_idx,
-                read_parent,
-                read_child,
-            ],
-        ),
+        extract_parent_rank,
+        extract_current_rank,
+        # Bubble Up
+        cb.while_with(current_lt_parent,
+                [
+                    cb.invoke(swap, in_a=parent_idx.out, in_b=current_idx.out, ref_mem=mem),
+                    set_idx_parent,
+                    find_parent_idx,
+                    extract_parent_rank
+                ])
+    ]
+
+    comp.control += [
+            cb.if_with(cmd_eq_0, 
+                cb.if_with(size_eq_0, raise_err, pop), 
+                cb.if_with(cmd_eq_1, 
+                    cb.if_with(size_eq_0, raise_err, peak),
+                    cb.if_with(cmd_eq_2, 
+                        cb.if_with(size_eq_max, raise_err, push), 
+                        raise_err
+                    )
+                )
+            )
     ]
 
     return comp
 
 
-def insert_main(prog, binheap):
-    """Inserts the main component into the program.
-    Invokes the `binheap` component with 64-bit values 4 and 5,
-    and a 64-bit memory of length 15.
+def insert_main(prog):
```

Comment:
Sweet. This is an excellent placeholder until we do the full monty with the oracles and the 20,000 values